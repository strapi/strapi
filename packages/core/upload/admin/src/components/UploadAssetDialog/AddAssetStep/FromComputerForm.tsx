import * as React from 'react';

import { useTracking } from '@strapi/admin/strapi-admin';
import { Box, Button, Flex, Modal, Typography } from '@strapi/design-system';
import { PlusCircle as PicturePlus } from '@strapi/icons';
import { useIntl } from 'react-intl';
import { styled } from 'styled-components';

import { AssetSource } from '../../../constants';
import { getTrad, rawFileToAsset } from '../../../utils';

import type { FileWithRawFile } from './AddAssetStep';

const Wrapper = styled(Flex)`
  flex-direction: column;
`;

const IconWrapper = styled.div`
  font-size: 6rem;

  svg path {
    fill: ${({ theme }) => theme.colors.primary600};
  }
`;

const MediaBox = styled(Box)`
  border-style: dashed;
`;

const OpaqueBox = styled(Box)`
  opacity: 0;
  cursor: pointer;
`;

interface FromComputerFormProps {
  onClose: () => void;
  onAddAssets: (assets: FileWithRawFile[]) => void;
  trackedLocation?: string;
}

export const FromComputerForm = ({
  onClose,
  onAddAssets,
  trackedLocation,
}: FromComputerFormProps) => {
  const { formatMessage } = useIntl();
  const [dragOver, setDragOver] = React.useState(false);
  const inputRef = React.useRef<HTMLInputElement>(null);
  const { trackUsage } = useTracking();

  const handleDragOver = (event: React.DragEvent<HTMLDivElement>) => {
    event.preventDefault();
  };

  const handleDragEnter = (event: React.DragEvent<HTMLDivElement>) => {
    event.preventDefault();
    setDragOver(true);
  };

  const handleDragLeave = () => setDragOver(false);

  const handleClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    // inputRef.current?.click();
  };

  const handleChange = () => {
    const files = inputRef.current?.files;
    const assets: FileWithRawFile[] = [];

    if (files) {
      for (let i = 0; i < files.length; i++) {
        const file = files.item(i);
        if (file) {
          const asset = rawFileToAsset(file, AssetSource.Computer);
          assets.push(asset);
        }
      }
    }

    if (trackedLocation) {
      trackUsage('didSelectFile', { source: 'computer', location: trackedLocation });
    }

    onAddAssets(assets);
  };

  const handleDrop = (e: React.DragEvent<HTMLDivElement>) => {
    e.preventDefault();

    if (e?.dataTransfer?.files) {
      const files = e.dataTransfer.files;
      const assets = [];

      for (let i = 0; i < files.length; i++) {
        const file = files.item(i);
        if (file) {
          const asset = rawFileToAsset(file, AssetSource.Computer);
          assets.push(asset);
        }
      }

      onAddAssets(assets);
    }

    setDragOver(false);
  };

  return (
    <form>
      <Box paddingLeft={8} paddingRight={8} paddingTop={6} paddingBottom={6}>
        <label>
          <MediaBox
            paddingTop={11}
            paddingBottom={11}
            hasRadius
            justifyContent="center"
            borderColor={dragOver ? 'primary500' : 'neutral300'}
            background={dragOver ? 'primary100' : 'neutral100'}
            position="relative"
            onDragEnter={handleDragEnter}
            onDragLeave={handleDragLeave}
            onDragOver={handleDragOver}
            onDrop={handleDrop}
          >
            <Flex justifyContent="center">
              <Wrapper>
                <IconWrapper>
                  <PicturePlus aria-hidden width="3.2rem" height="3.2rem" />
                </IconWrapper>

                <Box paddingTop={3} paddingBottom={5}>
                  <Typography variant="delta" textColor="neutral600" tag="span">
                    {formatMessage({
                      id: getTrad('input.label'),
                      defaultMessage: 'Drag & Drop here or',
                    })}
                  </Typography>
                </Box>

                <OpaqueBox
                  tag="input"
                  position="absolute"
                  left={0}
                  right={0}
                  bottom={0}
                  top={0}
                  width="100%"
                  type="file"
                  multiple
                  name="files"
                  aria-label={formatMessage({
                    id: getTrad('input.label'),
                    defaultMessage: 'Drag & Drop here or',
                  })}
                  tabIndex={-1}
                  ref={inputRef}
                  zIndex={1}
                  onChange={handleChange}
                />

                {/* <Box position="relative">
                  <Button type="button" onClick={handleClick}>
                    {formatMessage({
                      id: getTrad('input.button.label'),
                      defaultMessage: 'Browse files',
                    })}
                  </Button>
                </Box> */}
              </Wrapper>
            </Flex>
          </MediaBox>
        </label>
      </Box>

      <Modal.Footer>
        <Button onClick={onClose} variant="tertiary">
          {formatMessage({
            id: 'app.components.Button.cancel',
            defaultMessage: 'cancel',
          })}
        </Button>
      </Modal.Footer>
    </form>
  );
};
