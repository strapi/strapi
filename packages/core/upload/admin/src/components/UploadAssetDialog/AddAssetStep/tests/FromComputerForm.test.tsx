import { DesignSystemProvider } from '@strapi/design-system';
import { render as renderTL } from '@testing-library/react';
import { IntlProvider } from 'react-intl';

import { FromComputerForm } from '../FromComputerForm';

jest.mock('@strapi/admin/strapi-admin', () => ({
  ...jest.requireActual('@strapi/admin/strapi-admin'),
  getFetchClient: jest.fn().mockReturnValue({
    get: jest.fn(),
  }),
}));

describe('FromComputerForm', () => {
  it('snapshots the component', async () => {
    const { container } = renderTL(
      <IntlProvider locale="en" messages={{}}>
        <DesignSystemProvider>
          <FromComputerForm onClose={jest.fn()} onAddAssets={jest.fn()} />
        </DesignSystemProvider>
      </IntlProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
