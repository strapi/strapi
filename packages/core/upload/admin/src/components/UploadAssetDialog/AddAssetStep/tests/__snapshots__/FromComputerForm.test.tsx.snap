// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FromComputerForm snapshots the component 1`] = `
.c0 {
  padding-block-start: 24px;
  padding-block-end: 24px;
  padding-inline-start: 40px;
  padding-inline-end: 40px;
}

.c1 {
  padding-block-start: 64px;
  padding-block-end: 64px;
  background: #f6f6f9;
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  border-color: #c0c0cf;
  position: relative;
}

.c7 {
  padding-block-start: 12px;
  padding-block-end: 20px;
}

.c9 {
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  width: 100%;
  z-index: 1;
  position: absolute;
}

.c11 {
  padding-block-start: 16px;
  padding-inline-end: 20px;
  padding-block-end: 16px;
  padding-inline-start: 20px;
  background: #f6f6f9;
}

.c14 {
  padding-inline-start: 16px;
  padding-inline-end: 16px;
  border-radius: 4px;
  cursor: pointer;
}

.c3 {
  align-items: center;
  justify-content: center;
  flex-direction: row;
  display: flex;
}

.c4 {
  align-items: center;
  flex-direction: row;
  display: flex;
}

.c12 {
  align-items: center;
  justify-content: space-between;
  flex-direction: row;
  display: flex;
}

.c15 {
  gap: 8px;
  align-items: center;
  justify-content: center;
  flex-direction: row;
  display: inline-flex;
}

.c8 {
  font-weight: 500;
  font-size: 1.6rem;
  line-height: 1.25;
  color: #666687;
}

.c17 {
  font-size: 1.2rem;
  line-height: 1.33;
  color: currentcolor;
  font-weight: 600;
}

.c18 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c16 {
  height: 3.2rem;
  text-decoration: none;
  border: 1px solid #dcdce4;
  background: #ffffff;
  color: #32324d;
}

.c16:hover {
  background-color: #f6f6f9;
}

.c16:active {
  background-color: #eaeaef;
}

.c16[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
  color: #666687;
  cursor: default;
}

.c13 {
  border-top: solid 1px #eaeaef;
  flex: 1;
}

.c5 {
  flex-direction: column;
}

.c6 {
  font-size: 6rem;
}

.c6 svg path {
  fill: #4945ff;
}

.c2 {
  border-style: dashed;
}

.c10 {
  opacity: 0;
  cursor: pointer;
}

@media (prefers-reduced-motion: no-preference) {
  .c16 {
    transition: background-color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),border-color 200ms cubic-bezier(0.25, 0.46, 0.45, 0.94);
  }
}

<div>
  <form>
    <div
      class="c0"
    >
      <label>
        <div
          class="c1 c2"
          justifycontent="center"
        >
          <div
            class="c3"
          >
            <div
              class="c4 c5"
            >
              <div
                class="c6"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="3.2rem"
                  viewBox="0 0 32 32"
                  width="3.2rem"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M16 3a13 13 0 1 0 13 13A13.016 13.016 0 0 0 16 3m5 14h-4v4a1 1 0 0 1-2 0v-4h-4a1 1 0 0 1 0-2h4v-4a1 1 0 0 1 2 0v4h4a1 1 0 0 1 0 2"
                  />
                </svg>
              </div>
              <div
                class="c7"
              >
                <span
                  class="c8"
                >
                  Drag & Drop here or
                </span>
              </div>
              <input
                aria-label="Drag & Drop here or"
                class="c9 c10"
                multiple=""
                name="files"
                tabindex="-1"
                type="file"
              />
            </div>
          </div>
        </div>
      </label>
    </div>
    <footer
      class="c11 c12 c13"
    >
      <button
        aria-disabled="false"
        class="c14 c15 c16"
      >
        <span
          class="c17"
        >
          cancel
        </span>
      </button>
    </footer>
  </form>
  <span
    class="c18"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </span>
</div>
`;
